### PR TITLE
refactor(NODE-3405): implement MongoRuntimeError children

### PR DESCRIPTION
**Pull request by andymina**
## Description
Replace instances of MongoDriverError with the appropriate children of MongoRuntimeError, listed below, as described in docs/errors.md.

Children of MongoRuntimeError used:
- MongoTransactionError
- MongoExpiredSessionError
- MongoKerberosError
- MongoResourceClosedError
- MongoServerClosedError
- MongoStreamClosedError
- MongoTopologyClosedError

### CODE REVIEW

File: src/change_stream.ts [hunk 1]
```diff
@@ -1,32 +1,38 @@
...
+import type { ClientSession } from './sessions';
+import {
+  calculateDurationInMs,
+  Callback,
+  getTopology,
+  maxWireVersion,
+  maybePromise,
+  MongoDBNamespace,
+  now
+} from './utils';
```
Comment (W-A-James) [inline src/change_stream.ts:36]: Just out of curiosity, why the rearrangement of the imports?

File: src/utils.ts [hunk 2]
```diff
@@ -54,26 +54,31 @@ export function getSingleProperty(
  */
 export function checkCollectionName(collectionName: string): void {
   if ('string' !== typeof collectionName) {
+    // TODO(NODE-3405): Use MongoNamespace static method
```
Comment (W-A-James) [inline src/utils.ts:57]: We should probably make a ticket for this so we don't forget what exactly this refers to

File: src/cmap/connection_pool.ts [hunk 3]
```diff
@@ -320,7 +320,7 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
           clearTimeout(waitQueueMember.timer);
         }
         if (!waitQueueMember[kCancelled]) {
-          waitQueueMember.callback(new MongoDriverError('connection pool closed'));
+          waitQueueMember.callback(new MongoResourceClosedError('Connection pool closed'));
```
Comment (W-A-James) [inline src/cmap/connection_pool.ts:323]: Should we just go ahead and create a `MongoConnectionPoolClosedError`? It feels somewhat out of place having the more generic form of this error directly instantiated.

File: src/gridfs/upload.ts [hunk 4]
```diff
@@ -1,14 +1,20 @@
...
+import { PromiseProvider } from '../promise_provider';
 import type { Callback } from '../utils';
-import type { Document } from '../bson';
-import type { GridFSBucket } from './index';
-import type { GridFSFile } from './download';
 import type { WriteConcernOptions } from '../write_concern';
-import type { Collection } from '../collection';
+import { WriteConcern } from './../write_concern';
+import type { GridFSFile } from './download';
+import type { GridFSBucket } from './index';
```
Comment (W-A-James) [inline src/gridfs/upload.ts:17]: In the same vein as before, just curious as to what the intention was with reordering the imports.

File: src/sdam/server.ts [hunk 5]
```diff
@@ -62,6 +63,8 @@ const stateTransition = makeStateMachine({
   [STATE_CLOSING]: [STATE_CLOSING, STATE_CLOSED]
 });
 
+const SERVER_CLOSED_ERROR = 'Server is closed';
```
Comment (W-A-James) [inline src/sdam/server.ts:66]: In general, I think that when we have a resource that has been closed, the message should be of the form 

```javascript
'<resource type> has been closed'
```
That's a personal preference, but I think having it be more consistent would be better for error message readability. I'll flag other instances of this.

File: src/sdam/topology.ts [hunk 6]
```diff
@@ -83,6 +83,8 @@ const stateTransition = makeStateMachine({
   [STATE_CLOSING]: [STATE_CLOSING, STATE_CLOSED]
 });
 
+const TOPOLOGY_CLOSED_ERROR = 'Topology is closed';
```
Comment (W-A-James) [inline src/sdam/topology.ts:86]:
```suggestion
const TOPOLOGY_CLOSED_ERROR = 'Topology has been closed';
```

File: src/sdam/topology.ts [hunk 7]
```diff
@@ -956,7 +958,7 @@ function processWaitQueue(topology: Topology) {
   if (topology.s.state === STATE_CLOSED) {
     drainWaitQueue(
       topology[kWaitQueue],
-      new MongoDriverError('Topology is closed, please connect')
+      new MongoTopologyClosedError(TOPOLOGY_CLOSED_ERROR + ', please connect')
```
Comment (W-A-James) [inline src/sdam/topology.ts:961]: Change to use a template string

File: src/change_stream.ts [hunk 8]
```diff
@@ -678,13 +684,17 @@ function processNewChange<TSchema>(
   callback?: Callback<ChangeStreamDocument<TSchema>>
 ) {
   if (changeStream[kClosed]) {
-    if (callback) callback(new MongoDriverError(CHANGESTREAM_CLOSED_ERROR));
+    if (callback) callback(new MongoStreamClosedError(CHANGESTREAM_CLOSED_ERROR));
```
Comment (W-A-James) [inline src/change_stream.ts:687]: Tagging this here, but line 60 should read

```javascript
const CHANGESTREAM_CLOSED_ERROR = 'ChangeStream has been closed';
```

**Review by W-A-James — CHANGES_REQUESTED** [review]

File: src/change_stream.ts (hunk 1 above)
Comment (andymina) [reply src/change_stream.ts:36]: ESLint felt very strongly about having them in this order for some reason, lol.

File: src/sdam/server.ts (hunk 5 above)
Comment (andymina) [reply src/sdam/server.ts:66]: I'm a little hesitant to use the past tense in this error message because it might imply that the server has been closed as a result of this operation rather than conveying the current state of the server. Happy to see what everyone else thinks; maybe I'm being a little _too_ semantic.

File: src/cmap/connection_pool.ts (hunk 3 above)
Comment (andymina) [reply src/cmap/connection_pool.ts:323]: Sounds good! In general, I think it might be worth to make specific ClosedErrors for things that users will commonly run into (servers, streams, etc) and leaving `MongoResourceClosedError` as the general use case.

Comment (W-A-James) [reply src/cmap/connection_pool.ts:323]: In that case, this might be worth discussing with the team to get a better idea of the more common cases

File: src/gridfs/upload.ts (hunk 4 above)
Comment (andymina) [reply src/gridfs/upload.ts:17]: The almighty ESLint reordered things here too.

File: src/gridfs/upload.ts [hunk 9]
```diff
@@ -556,7 +562,7 @@ function writeRemnant(stream: GridFSBucketWriteStream, callback?: Callback): boo
 function checkAborted(stream: GridFSBucketWriteStream, callback?: Callback<void>): boolean {
   if (stream.state.aborted) {
     if (typeof callback === 'function') {
-      callback(new MongoDriverError('this stream has been aborted'));
+      callback(new MongoStreamClosedError('Stream has been aborted'));
```
Comment (W-A-James) [inline src/gridfs/upload.ts:565]: In the same vein as [this comment on NODE-3421](https://github.com/mongodb/node-mongodb-native/pull/2910#pullrequestreview-712206125) we may want to consider adding a default message for MongoStreamClosedError with an optional message parameter for cases where more information is needed

File: src/cmap/commands.ts [hunk 10]
```diff
@@ -84,8 +84,8 @@ export class Query {
 
     // Validate that we are not passing 0x00 in the collection name
     if (ns.indexOf('\x00') !== -1) {
-      // TODO(NODE-3483): Replace with MongoCommandError
-      throw new MongoDriverError('Namespace cannot contain a null character');
+      // TODO(NODE-3483): Use MongoNamespace static method
+      throw new MongoDriverError('namespace cannot contain a null character');
```
Comment (W-A-James) [inline src/cmap/commands.ts:88]: Why the change to start with a common letter here?

Comment (andymina) [reply src/cmap/commands.ts]: Oops, this got messed up in the rebase.

**Comment by andymina** [comment]
> @andymina There is a TODO that points to this ticket in `src/db.ts: 752`. Feel free to point that to [NODE-3484](https://jira.mongodb.org/browse/NODE-3484) instead

I removed the TODOs here because I realized this is the `validateDatabaseName()` mentioned in [NODE-3483](https://jira.mongodb.org/browse/NODE-3483). I left a TODO pointing to NODE-3483 for this entire function.

**Review by W-A-James — APPROVED** [review]

File: src/change_stream.ts [hunk 11]
```diff
@@ -1,5 +1,11 @@
 import Denque = require('denque');
-import { MongoError, AnyError, isResumableError, MongoDriverError } from './error';
+import {
+  MongoError,
+  AnyError,
+  isResumableError,
+  MongoDriverError,
+  MongoStreamClosedError
```
Comment (dariakp) [inline src/change_stream.ts:7]: As a general question to the team, do we want to conflate generic streams with change streams in our error reporting? Our change streams aren't really node streams in the sense that they don't extend from any of the base node streams. I could go either way, but my thought is that there could be some merit in differentiating "steam closed" vs "change stream closed" @nbbeeken @emadum Thoughts?

File: src/operations/execute_operation.ts [hunk 12]
```diff
@@ -87,8 +89,7 @@ export function executeOperation<
         owner = Symbol();
         session = topology.startSession({ owner, explicit: false });
       } else if (session.hasEnded) {
-        // TODO(NODE-3405): Change this out for MongoExpiredSessionError
-        return cb(new MongoDriverError('Use of expired sessions is not permitted'));
+        return cb(new MongoExpiredSessionError('Use of expired sessions is not permitted'));
       } else if (session.snapshotEnabled && !topology.capabilities.supportsSnapshotReads) {
         return cb(new MongoDriverError('Snapshot reads require MongoDB 5.0 or later'));
```
Comment (dariakp) [inline src/operations/execute_operation.ts:94]: I think this might have been missed as an api error because snapshots were implemented while the other PR was being implemented, can we add the correct compatibility error since we are touching this file?

File: src/sessions.ts [hunk 13]
```diff
@@ -286,12 +288,12 @@ export class ClientSession extends TypedEventEmitter<ClientSessionEvents> {
    */
   startTransaction(options?: TransactionOptions): void {
     if (this[kSnapshotEnabled]) {
-      throw new MongoDriverError('Transactions are not allowed with snapshot sessions');
+      throw new MongoTransactionError('Transactions are not allowed with snapshot sessions');
```
Comment (dariakp) [inline src/sessions.ts:291]: this is more of an API error - probably the compatibility one, because in order to get here, the user needs to intentionally start a snapshot session and then try to start a transaction on it

File: src/sessions.ts [hunk 14]
```diff
@@ -286,12 +288,12 @@ export class ClientSession extends TypedEventEmitter<ClientSessionEvents> {
...
     }
 
     assertAlive(this);
     if (this.inTransaction()) {
-      throw new MongoDriverError('Transaction already in progress');
+      throw new MongoTransactionError('Transaction already in progress');
```
Comment (dariakp) [inline src/sessions.ts:296]: This is also kind of an api error; we may need to rethink how we draw the line between runtime and api errors, specifically with respect to transactions: like, it's true that the driver is getting to an invalid state, but it's not "unexpected" because it's still the user making a choice; maybe the correct thing to do is subclass the transaction errors from API errors, or to subclass runtime errors into incorrect usage ones vs unexpected driver state, and then subclass these transaction errors according to the cause of the error

**Review by dariakp — CHANGES_REQUESTED** [review]
I didn't leave a comment on every line regarding transaction errors, but see my comment regarding reconsidering how we subclass these so that we can differentiate between incorrect usage from the user's side vs unexpected driver state

File: src/change_stream.ts (hunk 11 above)
Comment (nbbeeken) [reply src/change_stream.ts:7]: Seems worthy of differentiating, I would imagine we'd only need use of MongoChangeStreamError, but there may be instances of Streams related errors, unless they stop something from continuing in our code I would say node stream errors should be bubbled up as is.

Comment (andymina) [reply src/change_stream.ts:7]: The initial goal we had was to use `MongoStreamClosedError` between GridFSStream and ChangeStream to prevent creating two similar errors. If it's worth differentiating we could just split them down into `GridFSStreamClosed` and `ChangeStreamClosed`.

File: src/sessions.ts [hunk 15]
```diff
@@ -286,12 +289,12 @@ export class ClientSession extends TypedEventEmitter<ClientSessionEvents> {
...
   startTransaction(options?: TransactionOptions): void {
     if (this[kSnapshotEnabled]) {
-      throw new MongoDriverError('Transactions are not allowed with snapshot sessions');
+      throw new MongoCompatibilityError('Transactions are not allowed with snapshot sessions');
     }
 
     assertAlive(this);
     if (this.inTransaction()) {
-      throw new MongoDriverError('Transaction already in progress');
+      throw new MongoAPIError('Transaction already in progress');
```
Comment (W-A-James) [inline src/sessions.ts:297]: Since we are moving MongoTransactionError to be child of MongoAPIError, should we change this to MongoTransactionError and just go ahead and move MongoTransactionError to be a child of MongoAPIError in this PR?

**Review by W-A-James — CHANGES_REQUESTED** [review]

File: src/cmap/connection_pool.ts [hunk 16]
```diff
@@ -388,7 +388,7 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
           clearTimeout(waitQueueMember.timer);
         }
         if (!waitQueueMember[kCancelled]) {
-          waitQueueMember.callback(new MongoDriverError('connection pool closed'));
+          waitQueueMember.callback(new MongoAPIError('Connection pool closed'));
```
Comment (W-A-James) [inline src/cmap/connection_pool.ts:391]: Tag this with a comment for replacement in NODE-3485

File: src/sessions.ts (hunk 15 above)
Comment (W-A-James) [reply src/sessions.ts:381]: If not, we should also mark this for replacement in NODE-3485

**Review by W-A-James — APPROVED** [review]
LGTM after addressing the remaining comments

**Review by W-A-James — APPROVED** [review]

File: src/gridfs/upload.ts [hunk 17]
```diff
@@ -1,20 +1,13 @@
...
+import { MongoError, AnyError, MongoAPIError, MONGODB_ERROR_CODES } from '../error';
 import { WriteConcern } from './../write_concern';
-import type { GridFSFile } from './download';
+import { ObjectId } from '../bson';
+import { Callback, maybePromise } from '../utils';
+import type { Document } from '../bson';
 import type { GridFSBucket } from './index';
+import type { GridFSFile } from './download';
+import type { WriteConcernOptions } from '../write_concern';
+import type { Collection } from '../collection';
```
Comment (nbbeeken) [inline src/gridfs/upload.ts:10]: revert the import sorting here

File: src/sdam/topology.ts [hunk 18]
```diff
@@ -981,7 +988,7 @@ function processWaitQueue(topology: Topology) {
   if (topology.s.state === STATE_CLOSED) {
     drainWaitQueue(
       topology[kWaitQueue],
-      new MongoDriverError('Topology is closed, please connect')
+      new MongoTopologyClosedError(`${TOPOLOGY_CLOSED_ERROR}, please connect`)
```
Comment (nbbeeken) [inline src/sdam/topology.ts:991]: I think we can leave this error to be without message like the usage above

File: src/error.ts [hunk 19]
```diff
@@ -409,7 +409,7 @@ export class MongoCursorInUseError extends MongoAPIError {
  * @category Error
  */
 export class MongoServerClosedError extends MongoAPIError {
-  constructor(message: string) {
+  constructor(message = 'Server has been closed') {
```
Comment (nbbeeken) [inline src/error.ts:412]: Are we all in agreement on tense? super bikeshedding but present/imperative seems more ideal

File: src/sessions.ts (hunk 14 above)
Comment (andymina) [reply src/sessions.ts:380]: Capturing the rebucketing that happened in NODE-3421 where errors have been switched between MongoRuntimeError and MongoAPIError, with MongoTransactionError being one of them. MongoTransactionError is now a child of MongoAPIError.

File: src/sessions.ts [hunk 20]
```diff
@@ -441,6 +443,7 @@ export class ClientSession extends TypedEventEmitter<ClientSessionEvents> {
    * This is here to ensure that ClientSession is never serialized to BSON.
    */
   toBSON(): never {
+    // TODO(NODE-3484): Replace with MongoBSONParseError
```
Comment (nbbeeken) [inline src/sessions.ts:446]: This is one of those unreachables, its not a BSON parse error its really any error just to stop BSON from attempting to serialize a session, fair to leave as a driver error or runtime.

File: src/utils.ts [hunk 21]
```diff
@@ -258,8 +262,7 @@ export function executeLegacyOperation(
       const optionsIndex = args.length - 2;
       args[optionsIndex] = Object.assign({}, args[optionsIndex], { session: session });
     } else if (opOptions.session && opOptions.session.hasEnded) {
-      // TODO(NODE-3405): Replace this with MongoExpiredSessionError
-      throw new MongoDriverError('Use of expired sessions is not permitted');
+      throw new MongoExpiredSessionError('Use of expired sessions is not permitted');
```
Comment (nbbeeken) [inline src/utils.ts:265]: Let's make this error message the same has the one in applySession in sessions.ts
(maybe even make that the default message for expired session errors.)

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/error.ts (hunk 19 above)
Comment (andymina) [reply src/error.ts:412]: @dariakp @emadum @W-A-James? Personally, I'm in favor of present tense but happy to defer.

File: src/sessions.ts (hunk 20 above)
Comment (dariakp) [reply src/sessions.ts:447]: The TODO can be removed

File: src/error.ts (hunk 19 above)
Comment (emadum) [reply src/error.ts:412]: My vote would also be for using present/imperative.

File: src/operations/common_functions.ts [hunk 22]
```diff
@@ -42,7 +42,7 @@ export function indexInformation(
 
   // Did the user destroy the topology
   if (getTopology(db).isDestroyed())
-    return callback(new MongoDriverError('topology was destroyed'));
+    return callback(new MongoTopologyClosedError('Topology was destroyed'));
```
Comment (emadum) [inline src/operations/common_functions.ts:45]:
```suggestion
    return callback(new MongoTopologyClosedError('Topology is destroyed'));
```
If we're going to update messages to present tense, can do this one as well.

**Review by emadum — APPROVED** [review]
LGTM 👍

**Comment by andymina** [comment]
@dariakp re: [this comment](https://github.com/mongodb/node-mongodb-native/pull/2913#discussion_r686911454) You're totally right. I'm going to leave a TODO pointing toward [NODE-3483](https://jira.mongodb.org/browse/NODE-3483) which introduces MongoConnectionPoolClosedError as a child of RuntimeError. Also going to revert this to MongoRuntimeError for the time being.

**Review by nbbeeken — APPROVED** [review]

File: src/change_stream.ts [hunk 23]
```diff
@@ -683,15 +683,15 @@ function processNewChange<TSchema>(
...
+    if (callback) callback(new MongoAPIError(CHANGESTREAM_CLOSED_ERROR));
     return;
   }
 
   // a null change means the cursor has been notified, implicitly closing the change stream
   if (change == null) {
-    // TODO(NODE-3405): Replace with MongoStreamClosedError
-    return closeWithError(changeStream, new MongoDriverError(CHANGESTREAM_CLOSED_ERROR), callback);
+    // TODO(NODE-3485): Replace with MongoChangeStreamStreamClosedError
+    return closeWithError(changeStream, new MongoAPIError(CHANGESTREAM_CLOSED_ERROR), callback);
```
[the remark this answers is not in the document]
Comment (andymina) [reply src/change_stream.ts:695]: Yes, they're all API errors. They catch the case where someone has closed their ChangeStream before we are able to execute the logic beneath these checks. As an example, processing the resume queue will fail with the API error because the ChangeStream has been closed. So it does seem like incorrect usage.

File: src/change_stream.ts [hunk 24]
```diff
@@ -683,15 +683,15 @@ function processNewChange<TSchema>(
   callback?: Callback<ChangeStreamDocument<TSchema>>
 ) {
   if (changeStream[kClosed]) {
-    // TODO(NODE-3405): Replace with MongoStreamClosedError
-    if (callback) callback(new MongoDriverError(CHANGESTREAM_CLOSED_ERROR));
+    // TODO(NODE-3485): Replace with MongoChangeStreamStreamClosedError
```
[the remark this answers is not in the document]
Comment (dariakp) [reply src/change_stream.ts]: Can you update the other `StreamStream` comments in the file to match the spelling?

File: src/change_stream.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/change_stream.ts:695]: Sorry Andy I overlooked this one when we were on the call, you're correct. Except for this one, this is the one case where the `change` variable is null because the underlying cursor gave us null. Since change streams are "forever" this would only happen if there was some server reason for shutting the change stream so this would not be the fault of usage but rather a runtime issue.

[the remark this answers is not in the document]
Comment (dariakp) [reply src/change_stream.ts:695]: This specific one isn't an api error, also, could update all the TODO comments in this file to remove the duplicate `StreamStream` spelling?

[the remark this answers is not in the document]
Comment (andymina) [reply src/change_stream.ts:695]: Oops! Going to change this to be MongoRuntimeError and leave it as such. I think creating an entirely new error for this might be overkill.

**Review by dariakp — CHANGES_REQUESTED** [review]

**Review by dariakp — APPROVED** [review]

**Review by W-A-James — APPROVED** [review]
